DIRHASH_STANDARD.md

```diff
@@ -0,0 +1,203 @@
+- *Same collision resistance* as the underlying hashing algorithm.
+
+The process of computing the `DIRHASH` relies on the concepts below. Each concept is configurable by one or several options, which *affects the obtained hash value* (except for "Implementation" options):
+1. **Hash Function**: The underlying hash function used to map any data to a fixed length hash value.
+2. **Filtering**: The process of selecting what *entries* (subdirectories and files) within the directory to include.
+3. **Protocol**: Defines which data and metadata to include for each entry and precisely how it is fed to the hash function for a reproducible and collision resistant result.
+4. **Implementation**: Additional, implementation specific, aspects of how the `DIRHASH` is programmatically obtained.
+
+Since the configuration options for 1-3 above affects the obtained `DIRHASH`, they must be represented in a [`DIRHASH` based checksum](#the-dirsum-object) of a directory. Subsequently, the naming and interpretation of these options must be standardized to allow for communication and verification of such checksums.
+It is not required that a given package/module, implementing the `DIRHASH` standard, strictly follows the option naming (to adhere to language-specific convention e.g.) as long as they are properly translated to the standard options.
+
+Implementation options are naturally not covered by the Dirhash Standard as these *should not affect the obtained* `DIRHASH`. Such options should be clearly distinguished from those in 1-3.
+
+
+
+## Hash Function
+All data and metadata which forms the basis for the `DIRHASH` is hashed by a standard hash function. The Dirhash Standard defines precisely how data is fed to the hash function. The hash function is specified using the single option `algorithm`. The following algorithms are covered in the Dirhash Standard: [`"md5"`](https://www.ietf.org/rfc/rfc1321.txt), [`"sha1"`](https://tools.ietf.org/html/rfc3174), [`"sha224"`, `"sha256"`, `"sha384"`, `"sha512"`](https://tools.ietf.org/html/rfc6234) but can naturally be extended.
+
+
+### Hash function Options
+Name  | Type             | Default             | Description 
+----  | ---------------- | ------------------- | -----------
+algorithm | String | (no default) | The standard hashing algorithm (function) to use. 
+
+
+
+
+## Filtering
+Filtering governs what files and subdirectories to include. This is done by matching of file paths as well as handling of symbolic links and empty directories.
+
+
+### Filtering Options
+Name  | Type             | Default             | Description 
+----  | ---------------- | ------------------- | -----------
+match_patterns | Array of Strings | `["*"]` (match all) | Wildcard matching. Path relative to the Dirhash root is matched against the provided patterns. The path must match at least one of the "match patterns" (*not* starting with `!`) and none of the "ignore patterns" (starting with `!`).
```

Comment:
it's a glob right? https://en.wikipedia.org/wiki/Glob_(programming)

Reply:
file level or path level?

Reply:
i.e. will match on the full path or just a filename

Reply:
> it's a glob right? https://en.wikipedia.org/wiki/Glob_(programming)

Yeah, it's a glob, will add the link. Not sure exactly where the expected behaviour is formally specified, and I think there might be different "flavours"... In my implementation, I rely on https://github.com/cpburnz/python-path-specification 's `GitWildMatchPattern` So this might need further specification at some point.

Reply:
> file level or path level?

"Path relative to the Dirhash root is matched against the provided patterns" so path level.